x/ibc/07-tendermint/types/misbehaviour_handle.go

```diff
@@ -31,29 +31,28 @@ func (cs ClientState) CheckMisbehaviourAndUpdateState(
 	}
 
 	// If client is already frozen at earlier height than misbehaviour, return with error
-	height := clienttypes.NewHeight(0, misbehaviour.GetHeight())
-	if cs.IsFrozen() && cs.FrozenHeight.LTE(height) {
+	if cs.IsFrozen() && cs.FrozenHeight.LTE(misbehaviour.GetHeight()) {
 		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidMisbehaviour,
-			"client is already frozen at earlier height %d than misbehaviour height %d", cs.FrozenHeight, misbehaviour.GetHeight())
+			"client is already frozen at earlier height %s than misbehaviour height %s", cs.FrozenHeight, misbehaviour.GetHeight())
 	}
 
 	// Retrieve trusted consensus states for each Header in misbehaviour
 	// and unmarshal from clientStore
 
 	// Get consensus bytes from clientStore
-	tmConsensusState1, err := GetConsensusState(clientStore, cdc, tmMisbehaviour.Header1.TrustedHeight.EpochHeight)
+	tmConsensusState1, err := GetConsensusState(clientStore, cdc, tmMisbehaviour.Header1.TrustedHeight)
 	if err != nil {
-		return nil, sdkerrors.Wrapf(err, "could not get trusted consensus state from clientStore for Header1 at TrustedHeight: %d", tmMisbehaviour.Header1.TrustedHeight)
+		return nil, sdkerrors.Wrapf(err, "could not get trusted consensus state from clientStore for Header1 at TrustedHeight: %s", tmMisbehaviour.Header1)
 	}
 
 	// Get consensus bytes from clientStore
-	tmConsensusState2, err := GetConsensusState(clientStore, cdc, tmMisbehaviour.Header2.TrustedHeight.EpochHeight)
+	tmConsensusState2, err := GetConsensusState(clientStore, cdc, tmMisbehaviour.Header2.TrustedHeight)
 	if err != nil {
-		return nil, sdkerrors.Wrapf(err, "could not get trusted consensus state from clientStore for Header2 at TrustedHeight: %d", tmMisbehaviour.Header2.TrustedHeight)
+		return nil, sdkerrors.Wrapf(err, "could not get trusted consensus state from clientStore for Header2 at TrustedHeight: %s", tmMisbehaviour.Header2)
 	}
 
 	// calculate the age of the misbehaviour
-	infractionHeight := tmMisbehaviour.GetHeight()
+	infractionHeight := tmMisbehaviour.GetHeight().(clienttypes.Height).EpochHeight
```

Comment:
Hmm, I'm a little worried about this. Is it possible we could treat previously invalid evidence as valid again after an upgrade? I think we need to parse the epoch from the chain ID of the misbehaviour and check using the height comparator.